File: src/simulation/culture/aspect/Aspect.java
package simulation.culture.aspect;

import kotlin.Pair;
import simulation.Controller;
import simulation.culture.aspect.dependency.AspectDependencies;
import simulation.culture.aspect.dependency.Dependency;
import simulation.culture.group.centers.AspectCenter;
import simulation.culture.group.request.EvaluatorsKt;
import simulation.space.resource.MutableResourcePack;
import simulation.space.resource.Resource;
import simulation.space.resource.ResourcePack;
import simulation.space.resource.tag.ResourceTag;
import simulation.space.resource.tag.labeler.ResourceLabeler;
import simulation.space.resource.tag.labeler.TagLabeler;

import java.util.*;
import java.util.stream.Collectors;

public class Aspect {
    AspectCore aspectCore;
    /**
     * Map which stores for every requirement some Dependencies, from which
     * we can get get resource for using this aspect.
     */
    AspectDependencies dependencies = new AspectDependencies(new HashMap<>());
    /**
     * Coefficient which represents how much this aspect is used by its owner.
     */
    private int usefulness = 50;
    /**
     * Whether it was used on this turn.
     */
    public boolean used = false;
    private boolean usedThisTurn = false;
    public boolean canInsertMeaning = false;

    public Aspect(AspectCore aspectCore, AspectDependencies dependencies) {
        this.aspectCore = aspectCore;
        initDependencies(dependencies);
    }

    void initDependencies(AspectDependencies dependencies) {
        for (Map.Entry<ResourceTag, Set<Dependency>> entry : dependencies.getMap().entrySet()) {
            this.dependencies.getMap().put(
                    entry.getKey(),
                    entry.getValue().stream()
                            .map(Dependency::copy)
                            .collect(Collectors.toSet())
            );
        }
    }

    public void swapDependencies(AspectCenter aspectCenter) {
        dependencies.getMap().values().forEach(set ->
                set.forEach(d ->
                        d.swapDependencies(aspectCenter)
                )
        );
    }

    public String getName() {
        return aspectCore.getName();
    }

    public Collection<ResourceTag> getTags() {
        return aspectCore.getTags();
    }

    public Collection<ResourceTag> getRequirements() {
        return aspectCore.getRequirements();
    }

    public AspectDependencies getDependencies() {
        return dependencies;
    }

    public List<AspectMatcher> getMatchers() {
        return aspectCore.getMatchers();
    }

    public int getUsefulness() {
        return usefulness;
    }

    public boolean isValid() {
        return true;
    }

    public boolean canApplyMeaning() {
        return aspectCore.getApplyMeaning();
    }

    public boolean canReturnMeaning() {
        return this instanceof ConverseWrapper && ((ConverseWrapper) this).canInsertMeaning;
    }

    public boolean isDependenciesOk(AspectDependencies dependencies) {
        return getRequirements().size() == dependencies.getSize();
    }

    public boolean canTakeResources() {
        return false;
    }

    @Deprecated
    public void mergeDependencies(Aspect aspect) {//TODO what's going on here?
        if (!aspect.equals(this)) {
            return;
        }

        for (ResourceTag tag : dependencies.getMap().keySet()) {
            dependencies.getMap().get(tag).addAll(aspect.dependencies.getMap().get(tag));
        }
    }

    public void addOneDependency(AspectDependencies newDependencies) {
        for (ResourceTag tag : dependencies.getMap().keySet()) {
            try {
                for (Dependency dependency1 : newDependencies.getMap().get(tag)) {
                    if (!dependencies.getMap().get(tag).contains(dependency1)) {
                        dependencies.getMap().get(tag).add(dependency1);
                        break;
                    }
                }
            } catch (Exception e) {
                System.err.println(e);
                int i = 0;
            }
        }
    }

    public Aspect copy(AspectDependencies dependencies) {
        return aspectCore.copy(dependencies);
    }

    public List<Resource> getProducedResources() {
        return Collections.emptyList();
    }

    public AspectResult use(AspectController controller) {//TODO instrument efficiency
        //TODO put dependency resources only in node; otherwise they may merge with phony
        if (controller.getDepth() > Controller.session.maxGroupDependencyDepth || used) {
            return new AspectResult(
                    false,
                    new ArrayList<>(),
                    new MutableResourcePack(),
                    new AspectResult.ResultNode(this)
            );
        }
        used = true;
        boolean isFinished = true;
        List<Pair<ResourceLabeler, Integer>> neededResources = new ArrayList<>();
        MutableResourcePack meaningfulPack = new MutableResourcePack();
        int neededWorkers = controller.getCeilingSatisfiableAmount(getProducedResources());
        int gotWorkers = controller.getPopulationCenter().changeStratumAmountByAspect((ConverseWrapper) this, neededWorkers);
        controller.setMax(gotWorkers);
        AspectResult.ResultNode node = new AspectResult.ResultNode(this);
        if (controller.getCeiling() > 0) {
            for (Map.Entry<ResourceTag, Set<Dependency>> entry : dependencies.getNonPhony().entrySet()) {
                if (!satisfyRegularDependency(controller, entry.getKey(), entry.getValue(), meaningfulPack, node)) {
                    isFinished = false;
                    neededResources.add(new Pair<>(new TagLabeler(entry.getKey()), controller.getCeiling()));
                }
            }
        }
        if (isFinished) {
            isFinished = satisfyPhonyDependency(controller, dependencies.getPhony(), meaningfulPack);
        }
        if (controller.isFloorExceeded(meaningfulPack)) {
            markAsUsed();
        } else {
            controller.getPopulationCenter().freeStratumAmountByAspect((ConverseWrapper) this, gotWorkers);
        }
        used = false;
        return new AspectResult(isFinished, neededResources, meaningfulPack, node);
    }

    private boolean satisfyPhonyDependency(
            AspectController controller,
            Set<Dependency> dependencies,
            MutableResourcePack meaningfulPack
    ) {
        meaningfulPack.addAll(getPhonyFromResources(controller));
        if (controller.isCeilingExceeded(meaningfulPack)) {
            return true;
        }
        for (Dependency dependency : dependencies) {
            int newDelta = meaningfulPack.getAmount(((ConverseWrapper) this).resource);
            AspectResult _p = dependency.useDependency(new AspectController(
                    controller.getDepth() + 1,
                    controller.getCeiling() - newDelta,
                    controller.getFloor() - newDelta,
                    controller.getEvaluator(),
                    controller.getPopulationCenter(),
                    controller.getTerritory(),
                    shouldPassMeaningNeed(controller.isMeaningNeeded()),
                    controller.getMeaning()
            ));
            if (!_p.isFinished) {
                continue;
            }
            meaningfulPack.addAll(_p.resources);
            if (controller.isCeilingExceeded(meaningfulPack)) {
                break;
            }
        }
        return true;
    }

    ResourcePack getPhonyFromResources(AspectController controller) {
        ResourcePack pack = EvaluatorsKt.resourceEvaluator(((ConverseWrapper) this).resource).pick(
                controller.getPopulationCenter().getTurnResources()
        );
        return controller.pickCeilingPart(
                pack.getResources(),
                r -> r.applyAspect(((ConverseWrapper) this).aspect),
                (r, n) -> r.applyAndConsumeAspect(((ConverseWrapper) this).aspect, n)
        );
    }

    private boolean satisfyRegularDependency(
            AspectController controller,
            ResourceTag requirementTag,
            Set<Dependency> dependencies,
            MutableResourcePack meaningfulPack,
            AspectResult.ResultNode node
    ) {
        boolean isFinished = false;
        MutableResourcePack _rp = new MutableResourcePack();
        _rp.addAll(
                controller.pickCeilingPart(
                        controller.getPopulationCenter().getStratumByAspect((ConverseWrapper) this)
                                .getInstrumentByTag(requirementTag).getResources(),
                        r -> Collections.singletonList(r.copy(1)),
                        (r, n) -> Collections.singletonList(r.getCleanPart(n))
                )
        );
        MutableResourcePack usedForDependency = new MutableResourcePack();

        for (Dependency dependency : dependencies) {
            int newDelta = _rp.getAmount(requirementTag);
            AspectResult result = dependency.useDependency(new AspectController(
                    controller.getDepth() + 1,
                    controller.getCeiling() - newDelta,
                    controller.getFloor() - newDelta,
                    controller.getEvaluator(),
                    controller.getPopulationCenter(),
                    controller.getTerritory(),
                    false,
                    controller.getMeaning()
            ));
            _rp.addAll(result.resources);
            if (!result.isFinished) {
                continue;
            }
            if (_rp.getAmount(requirementTag) >= controller.getCeiling()) {
                if (!requirementTag.isInstrumental) {
                    //TODO sometimes can spend resources without getting result because other dependencies are lacking
                    usedForDependency.addAll(_rp.getAmountOfResourcesWithTagAndErase(
                            requirementTag,
                            controller.getCeiling()).getSecond()
                    );
                } else {
                    usedForDependency.addAll(_rp.getResources(requirementTag));
                }
                meaningfulPack.addAll(_rp);
                isFinished = true;
                break;
            }
        }
        node.resourceUsed.put(requirementTag, usedForDependency);
        return isFinished;
    }

    protected boolean shouldPassMeaningNeed(boolean isMeaningNeeded) {
        return isMeaningNeeded;
    }

    protected void markAsUsed() {
        usefulness += 1;
        usedThisTurn = true;
    }

    public void finishUpdate() {
        if (!usedThisTurn) {
            usefulness--;
        }
        usedThisTurn = false;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Aspect aspect = (Aspect) o;
        return aspectCore.getName().equals(aspect.aspectCore.getName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(aspectCore.getName());
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder("Aspect " + aspectCore.getName() + ", usefulness - " + usefulness +
                ", dependencies:");
        for (Map.Entry<ResourceTag, Set<Dependency>> entry : dependencies.getMap().entrySet()) {
            stringBuilder.append("\n**").append(entry.getKey().name).append(":");
            for (Dependency dependency : entry.getValue()) {
                stringBuilder.append("\n**").append(dependency.getName());
            }
        }
        return stringBuilder.toString();
    }
}

File: src/simulation/culture/aspect/Aspect.kt
package simulation.culture.aspect

import shmp.random.testProbability
import simulation.Controller.*
import simulation.SimulationError
import simulation.culture.aspect.AspectResult.ResultNode
import simulation.culture.aspect.dependency.AspectDependencies
import simulation.culture.aspect.dependency.Dependency
import simulation.culture.group.centers.AspectCenter
import simulation.culture.group.request.ResourceEvaluator
import simulation.culture.group.request.resourceEvaluator
import simulation.space.resource.container.MutableResourcePack
import simulation.space.resource.Resource
import simulation.space.resource.container.ResourcePack
import simulation.space.resource.tag.ResourceTag
import simulation.space.resource.tag.labeler.BaseNameLabeler
import simulation.space.resource.tag.labeler.ResourceLabeler
import simulation.space.resource.tag.labeler.TagLabeler
import java.util.*
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

open class Aspect(var core: AspectCore, dependencies: AspectDependencies) {
    /**
     * Map which stores for every requirement some Dependencies, from which
     * Aspect can get get resource for use.
     */
    var dependencies = AspectDependencies(mutableMapOf())

    init {
        initDependencies(dependencies)
    }

    var usefulness = session.defaultAspectUsefulness
        private set

    /**
     * Whether it was used on this turn.
     */
    var used = false
    private var usedThisTurn = false
    private var tooManyFailsThisTurn = false
    private var timesUsedInTurn = 0
    private var timesUsedInTurnUnsuccessfully = 0

    var canInsertMeaning = false

    fun initDependencies(dependencies: AspectDependencies) {
        for ((key, value) in dependencies.map)
            this.dependencies.map[key] = value.map { it.copy() }.toMutableSet()
    }

    open fun swapDependencies(aspectCenter: AspectCenter) =
            dependencies.map.values.forEach { set -> set.forEach { it.swapDependencies(aspectCenter) } }

    val name = core.name

    val tags = core.tags

    val requirements = core.requirements

    open val isValid = true

    fun canApplyMeaning() = core.applyMeaning

    fun canReturnMeaning() = this is ConverseWrapper && this.canInsertMeaning

    open fun isDependenciesOk(dependencies: AspectDependencies) = requirements.size == dependencies.size

    open fun canTakeResources() = false

    fun addOneDependency(newDependencies: AspectDependencies) {
        for (tag in dependencies.map.keys) try {
            for (dependency1 in newDependencies.map.getValue(tag)) {
                if (!dependencies.map.getValue(tag).contains(dependency1)) {
                    dependencies.map.getValue(tag).add(dependency1)
                    break
                }
            }
        } catch (e: Exception) {
            System.err.println(e)
            val i = 0
        }
    }

    open fun copy(dependencies: AspectDependencies) = core.makeAspect(dependencies)

    open val producedResources: List<Resource> = emptyList()

    fun calculateNeededWorkers(evaluator: ResourceEvaluator, amount: Double) = ceil(max(
            evaluator.getSatisfiableAmount(amount, producedResources) * core.standardComplexity,
            1.0
    )).toInt()

    fun calculateProducedValue(evaluator: ResourceEvaluator, workers: Int) =
            (evaluator.evaluate(producedResources) * workers) / core.standardComplexity

    private fun checkTermination(controller: AspectController) = tooManyFailsThisTurn
            || controller.depth > session.maxGroupDependencyDepth
            || used
            || core.resourceExposed && producedResources.any {
        val center = controller.territory.center
                ?: throw SimulationError("Empty Territory for the Aspect use")
        !it.isAcceptable(center)
    }

    protected fun _use(controller: AspectController): AspectResult {
        //TODO put dependency resources only in node; otherwise they may merge with phony
        if (checkTermination(controller)) return AspectResult(
                false,
                ArrayList(),
                MutableResourcePack(),
                ResultNode(this)
        )

        this as ConverseWrapper
        timesUsedInTurn++
        used = true
        var isFinished = true
        val neededResources = mutableListOf<Pair<ResourceLabeler, Int>>()
        val meaningfulPack = MutableResourcePack()

        val neededWorkers = calculateNeededWorkers(controller.evaluator, controller.ceiling)
        val gotWorkers = controller.populationCenter.getPeopleByAspect(this, neededWorkers)
        val allowedAmount = min(
                gotWorkers.cumulativeWorkers / core.standardComplexity
                        * controller.evaluator.evaluate(producedResources),
                controller.ceiling
        )

        controller.setMax(allowedAmount)
        val node = ResultNode(this)

        if (controller.ceiling > 0)
            for ((key, value) in dependencies.nonPhony.entries) {
                val (isOk, needs) = satisfyRegularDependency(controller, key, value, meaningfulPack, node)
                neededResources.addAll(needs)
                if (!isOk) {
                    isFinished = false
                    neededResources.add(TagLabeler(key) to ceil(controller.ceiling).toInt())
                }
            }

        if (isFinished)
            isFinished = satisfyPhonyDependency(controller, dependencies.phony, meaningfulPack)
        if (controller.isFloorExceeded(meaningfulPack))
            markAsUsed()
        else {
            controller.populationCenter.freeStratumAmountByAspect(this, gotWorkers)
            val neededAmount = ceil(controller.floor - controller.evaluate(meaningfulPack)).toInt()
            neededResources.add(BaseNameLabeler(resource.baseName) to neededAmount)
        }
        if (!isFinished) {
            timesUsedInTurnUnsuccessfully++
            if (!testProbability(1.0 / timesUsedInTurnUnsuccessfully.toDouble().pow(0.05), session.random))
                tooManyFailsThisTurn = true
        }
        used = false

        return AspectResult(isFinished, neededResources, meaningfulPack, node)
    }

    private fun satisfyPhonyDependency(
            controller: AspectController,
            dependencies: Set<Dependency>,
            meaningfulPack: MutableResourcePack
    ): Boolean {
        meaningfulPack.addAll(getPhonyFromResources(controller))
        if (controller.isCeilingExceeded(meaningfulPack))
            return true
        for (dependency in dependencies) {
            val newDelta = meaningfulPack.getAmount((this as ConverseWrapper).resource)
            val _p = dependency.useDependency(controller.copy(
                    depth = controller.depth + 1,
                    ceiling = controller.ceiling - newDelta,
                    floor = controller.floor - newDelta,
                    isMeaningNeeded = shouldPassMeaningNeed(controller.isMeaningNeeded)
            ))
            if (!_p.isFinished)
                continue
            meaningfulPack.addAll(_p.resources)
            if (controller.isCeilingExceeded(meaningfulPack))
                break
        }
        return true
    }

    private fun getPhonyFromResources(controller: AspectController): ResourcePack {
        val pack = resourceEvaluator((this as ConverseWrapper).resource).pick(controller.populationCenter.turnResources)
        return controller.pickCeilingPart(
                pack.resources,
                { it.applyAction(aspect.core.resourceAction) }
        ) { r, n -> r.applyActionAndConsume(aspect.core.resourceAction, n, true) }
    }

    private fun satisfyRegularDependency(
            controller: AspectController,
            requirementTag: ResourceTag,
            dependencies: Set<Dependency>,
            meaningfulPack: MutableResourcePack,
            node: ResultNode
    ): Result {
        val needs = mutableListOf<Need>()
        var isFinished = false
        val _rp = MutableResourcePack()
        _rp.addAll(controller.pickCeilingPart(
                controller.populationCenter.stratumCenter.getByAspect(this as ConverseWrapper)
                        .getInstrumentByTag(requirementTag).resources,
                { listOf(it.copy(1)) }
        ) { r, n -> listOf(r.getCleanPart(n)) })
        val usedForDependency = MutableResourcePack()
        for (dependency in dependencies) {
            val newDelta = _rp.getAmount(requirementTag)
            val result = dependency.useDependency(controller.copy(
                    depth = controller.depth + 1,
                    ceiling = controller.ceiling - newDelta,
                    floor = controller.floor - newDelta,
                    isMeaningNeeded = false
            ))
            needs.addAll(result.neededResources)
            _rp.addAll(result.resources)
            if (!result.isFinished)
                continue
            if (_rp.getAmount(requirementTag) >= controller.ceiling) {
                //TODO sometimes can spend resources without getting result because other dependencies are lacking
                if (!requirementTag.isInstrumental)
                    usedForDependency.addAll(_rp.getAmountOfResourcesWithTagAndErase(
                            requirementTag,
                            controller.ceiling
                    ).second)
                else
                    usedForDependency.addAll(_rp.getResources(requirementTag))
                meaningfulPack.addAll(_rp)
                isFinished = true
                break
            } else {
                val j = 0
            }
        }
        node.resourceUsed[requirementTag] = usedForDependency
        return Result(isFinished, needs)
    }

    protected open fun shouldPassMeaningNeed(isMeaningNeeded: Boolean) = isMeaningNeeded

    fun markAsUsed() = gainUsefulness(1)

    fun gainUsefulness(amount: Int) {
        if (amount <= 0) return
        usefulness = max(usefulness + amount, session.defaultAspectUsefulness)
        usedThisTurn = true
    }

    fun finishUpdate() {
        timesUsedInTurn = 0
        timesUsedInTurnUnsuccessfully = 0
        tooManyFailsThisTurn = false
        if (!usedThisTurn)
            usefulness--
        usedThisTurn = false
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true

        if (other == null || javaClass != other.javaClass)
            return false

        val aspect = other as Aspect
        return core.name == aspect.core.name
    }

    override fun hashCode() = Objects.hash(core.name)

    override fun toString() = "Aspect ${core.name} usefulness - $usefulness dependencies:\n" +
            dependencies.map.entries.joinToString("\n") { (tag, deps) ->
                "**${tag.name}:\n" +
                        deps.joinToString("\n") { "****${it.name}" }
            }
}

private data class Result(val isFinished: Boolean = true, val need: List<Need> = emptyList())

typealias Need = Pair<ResourceLabeler, Int>
